lib/bundler/plugin/index.rb

```
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+require "yaml"
+
+module Bundler
+  # Manages which plugins are installed and their sources. This also is supposed to map
+  # which plugin does what (currently the features are not implemented so this class is
+  # now a stub class).
+  class Plugin::Index
+    def initialize
+      @plugin_paths = {}
+      @commands = {}
+
+      load_index
+    end
+
+    # This function is to be called when a new plugin is installed. This function shall add
+    # the functions of the plugin to existing maps and also the name to source location.
+    #
+    # @param [String] name of the plugin to be registered
+    # @param [String] path where the plugin is installed
+    # @param [Array<String>] commands that are handled by the plugin
+    def register_plugin(name, path, commands)
+      @plugin_paths[name] = path
+
+      common = commands & @commands.keys
+      raise "Command(s) #{common.join(", ")} are already registered" if common.any?
```

this should raise a specific error